### PR TITLE
[Core] C API for Manager

### LINKED ISSUE #357: C API for `Manager`
## What

Add C API to the host API `Manager` class. 
Continues from #345.

## ACs

- A C handle based API (no need for a suite as this is a final API class)
  - Include constructor that takes a C handle to a ManagerInterface
        - `ManagerInterface` handle needs to changed to be a shared ptr (and existing tests updated) [separate PR]
  - `Manager` handle can be a raw ptr owned by the caller of the ctor
- C API tests - calling a C manager to a C++ manager interface
   - Write mock C++ manager interface implementation (shared test resource)
## Out of scope
- No `Session` in C

### PR DESCRIPTION
**Pull request by feltech**
Closes #357

Adds the `Manager` C API. Hence also refactors some of the opaque handle utilities and naming.

### CODE REVIEW

File: tests/openassetio-core-c/private/hostAPI/ManagerTest.cpp [hunk 1]
```diff
@@ -246,3 +246,69 @@ SCENARIO("A host calls Manager::displayName") {
...
+
+        // C API call.
+        OPENASSETIO_NS(ErrorCode)
+        code =
+            OPENASSETIO_NS(hostAPI_Manager_info)(&actualErrorMsg, actualInfoHandle, managerHandle);
+
+        THEN("generic exception error code and message is set and info is unmodified") {
+          CHECK(code == OPENASSETIO_NS(ErrorCode_kException));
+          CHECK(actualErrorMsg == expectedErrorMsg);
+          CHECK(actualInfo.empty());
```
Comment (foundrytom) [inline tests/openassetio-core-c/private/hostAPI/ManagerTest.cpp:298]: [consider] Same point about non-default initial values.

**Review by foundrytom — CHANGES_REQUESTED** [review]
WIP review

File: src/openassetio-core-c/private/handles/managerAPI/ManagerInterface.hpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+#include <openassetio/export.h>
+
+#include <openassetio/managerAPI/ManagerInterface.hpp>
+
+#include "../Converter.hpp"
+
+namespace openassetio {
+inline namespace OPENASSETIO_VERSION {
+namespace handles::managerAPI {
+using ManagerInterface = Converter<openassetio::managerAPI::ManagerInterfacePtr,
```
Comment (foundrytom) [inline src/openassetio-core-c/private/handles/managerAPI/ManagerInterface.hpp:15]: [question] Should we use a naming convention that differentiates shared ownership handles from the others?

**Review by foundrytom — COMMENTED** [review]
The naming/structural improvements are great @feltech.

File: src/openassetio-core-c/include/openassetio/c/hostAPI/Manager.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+ * Constructor function.
+ *
+ * Allocates a new \fqref{hostAPI::Manager} "Manager", which should be
+ * deallocated by \fqcref{hostAPI_Manager_dtor} "dtor" when the
+ * `Manager` is no longer in use.
+ *
+ * @param[out] err Storage for error message, if any.
+ * @param[out] handle
+ * @param managerInterfaceHandle  A handle to a
+ * \fqref{managerAPI::ManagerInterface} that the `Manager` will delegate
```
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core-c/include/openassetio/c/hostAPI/Manager.h:38]: I'm not sure. In this particular case the `ManagerInterface` handle is a handle to a `shared_ptr`, so the host is free to (in fact should) `dtor` it. 

A good point to be made more generally about calling out whether ownership is transferred or not when passing handles around.

File: tests/openassetio-core-c/private/hostAPI/ManagerTest.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      // Construct Manager through C API.
+      OPENASSETIO_NS(ErrorCode)
+      actualErrorCode = OPENASSETIO_NS(hostAPI_Manager_ctor)(&actualErrorMsg, &managerHandle,
+                                                             mockManagerInterfaceHandle);
+      CHECK(actualErrorCode == OPENASSETIO_NS(ErrorCode_kOK));
+
+      WHEN("Manager's dtor C API function is called") {
+        OPENASSETIO_NS(hostAPI_Manager_dtor)(managerHandle);
+
+        THEN("wrapped ManagerInterface is not destroyed") {}
```
[the remark this answers is not in the document]
Comment (feltech) [reply tests/openassetio-core-c/private/hostAPI/ManagerTest.cpp]: This was already addressed in a subsequent commit (9b797d61).  I should really have amended the previous one.

File: src/openassetio-core-c/include/openassetio/c/hostAPI/Manager.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (foundrytom) [reply src/openassetio-core-c/include/openassetio/c/hostAPI/Manager.h:38]: Ah of course, sorry was being dense.

File: src/openassetio-core-c/private/hostAPI/Manager.cpp [hunk 5]
```diff
@@ -50,10 +49,10 @@ void OPENASSETIO_NS(hostAPI_Manager_dtor)(OPENASSETIO_NS(hostAPI_Manager_h) hand
...
 }
 
-OPENASSETIO_CORE_C_EXPORT OPENASSETIO_NS(ErrorCode)
-    OPENASSETIO_NS(hostAPI_Manager_identifier)(OPENASSETIO_NS(StringView) * err,
-                                               OPENASSETIO_NS(StringView) * out,
-                                               OPENASSETIO_NS(hostAPI_Manager_h) handle) {
+OPENASSETIO_NS(ErrorCode)
+OPENASSETIO_NS(hostAPI_Manager_identifier)
+(OPENASSETIO_NS(StringView) * err, OPENASSETIO_NS(StringView) * out,
+ OPENASSETIO_NS(hostAPI_Manager_h) handle) {
```
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core-c/private/hostAPI/Manager.cpp:49]: As discussed offline, yes it was `clang-format` and will likely look much neater once #370 is done.

File: src/openassetio-core-c/private/handles/managerAPI/ManagerInterface.hpp (hunk 2 above)
Comment (feltech) [reply src/openassetio-core-c/private/handles/managerAPI/ManagerInterface.hpp:15]: Hmm, how about we make it more consistent to become `using ManagerInterfacePtr = ...`? 

The only time we would explicitly suffix with `Ptr` would be the `SharedPtr` case.  Though at-a-glance a new user wouldn't necessarily make that connection.

`using ManagerInterfaceSharedPtr = ...`, then?

File: src/openassetio-core-c/include/openassetio/c/hostAPI/Manager.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core-c/include/openassetio/c/hostAPI/Manager.h:38]: No its a good point (just in the opposite sense), I'll add a comment.

File: src/openassetio-core-c/private/handles/managerAPI/ManagerInterface.hpp (hunk 2 above)
Comment (foundrytom) [reply src/openassetio-core-c/private/handles/managerAPI/ManagerInterface.hpp:15]: I think you're right about the suffix ambiguity for the "normal" case. How about
`using SharedManagerInterface = ...` as we're already in the `handles` space?

File: tests/openassetio-core-c/private/hostAPI/ManagerTest.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  GIVEN("a shared pointer to a ManagerInterface and its C handle") {
+    auto* managerInterface = new DeathwatchedMockManagerInterface{};
+    // We must have this expectation here to avoid a false positive.
+    REQUIRE_DESTRUCTION(*managerInterface);
+    // Wrap deathwatched mock ManagerInterface in a SharedPtr.
+    openassetio::managerAPI::ManagerInterfacePtr mockManagerInterfacePtr{managerInterface};
+    // Convert the ManagerInterface pointer to a handle.
+    auto* mockManagerInterfaceHandle =
```
[the remark this answers is not in the document]
Comment (feltech) [reply tests/openassetio-core-c/private/hostAPI/ManagerTest.cpp]: Removed a bunch of uses of auto in 1628df35 and 32143c88

File: tests/openassetio-core-c/private/hostAPI/ManagerTest.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  OPENASSETIO_NS(StringView) actualErrorMsg{errStorage.size(), errStorage.data(), 0};
+  // A mock ManagerInterface whose lifetime is tracked.
+  using DeathwatchedMockManagerInterface = trompeloeil::deathwatched<MockManagerInterface>;
+
+  GIVEN("a shared pointer to a ManagerInterface and its C handle") {
+    auto* managerInterface = new DeathwatchedMockManagerInterface{};
+    // Wrap deathwatched mock ManagerInterface in a SharedPtr.
+    openassetio::managerAPI::ManagerInterfacePtr mockManagerInterfacePtr{managerInterface};
+    // Convert the ManagerInterface pointer to a handle.
+    auto* mockManagerInterfaceHandle =
```
[the remark this answers is not in the document]
Comment (feltech) [reply tests/openassetio-core-c/private/hostAPI/ManagerTest.cpp]: Removed a bunch of uses of auto in 1628df35 and 32143c88

File: tests/openassetio-core-c/private/hostAPI/ManagerTest.cpp [hunk 8]
```diff
@@ -102,7 +113,135 @@ SCENARIO("A Manager is constructed and destructed") {
...
+
+        // C API call.
+        OPENASSETIO_NS(ErrorCode)
+        code = OPENASSETIO_NS(hostAPI_Manager_identifier)(&actualErrorMsg, &actualIdentifier,
+                                                          managerHandle);
+
+        THEN("generic exception error code and message is set and identifier is unmodified") {
+          CHECK(code == OPENASSETIO_NS(ErrorCode_kException));
+          CHECK(actualErrorMsg == expectedErrorMsg);
+          CHECK(actualIdentifier == "");
```
[the remark this answers is not in the document]
Comment (feltech) [reply tests/openassetio-core-c/private/hostAPI/ManagerTest.cpp]: Updated in 80802931

File: src/openassetio-core-c/private/hostAPI/Manager.cpp [hunk 9]
```diff
@@ -62,15 +61,33 @@ OPENASSETIO_CORE_C_EXPORT OPENASSETIO_NS(ErrorCode)
...
+OPENASSETIO_NS(hostAPI_Manager_info)
+(OPENASSETIO_NS(StringView) * err, OPENASSETIO_NS(InfoDictionary_h) out,
+ OPENASSETIO_NS(hostAPI_Manager_h) handle) {
+  return openassetio::errors::catchUnknownExceptionAsCode(err, [&] {
+    using InfoDictHandleConverter =
+        openassetio::handles::Converter<openassetio::InfoDictionary,
+                                        OPENASSETIO_NS(InfoDictionary_h)>;
+
+    auto* outDict = InfoDictHandleConverter::toInstance(out);
+    const auto* manager = ManagerHandleConverter::toInstance(handle);
```
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core-c/private/hostAPI/Manager.cpp]: I agree if it hurts readability then it shouldn't be used. 

Removed a bunch of uses of auto in 1628df35 and 32143c88

File: src/openassetio-core-c/private/handles/managerAPI/ManagerInterface.hpp (hunk 2 above)
Comment (feltech) [reply src/openassetio-core-c/private/handles/managerAPI/ManagerInterface.hpp]: Updated to your suggestion, both converter and handle, in 85e62d85.  I'm not sure about this, but then we need to explore more what the lifecycle of a `managerAPI_SharedManagerInterface_h` looks like, and if it even needs to exist.

**Review by foundrytom — APPROVED** [review]
Nice work @feltech, thanks for patching all those up. LGTMSAMAYLATW

**Pull request merged**